extensions/amp-story/0.1/amp-story.js

```diff
@@ -19,34 +19,734 @@
+    const pagesByDistance = [];
+    Object.keys(distanceMap).forEach(pageId => {
+      const distance = distanceMap[pageId];
+      if (!pagesByDistance[distance]) {
+        pagesByDistance[distance] = [];
+      }
+      pagesByDistance[distance].push(pageId);
+    });
+
+    return pagesByDistance;
+  }
+
+  /**
+   * Creates a map of a page and all of the pages reachable from that page, by
+   * distance.
+   *
+   * @param {number} distance The distance that the page with the specified
+   *     pageId is from the active page.
+   * @param {!Object<string, number>} map A mapping from pageId to its distance
+   *     from the active page.
+   * @param {string} pageId The page to be added to the map.
+   * @return {!Object<string, number>} A mapping from page ID to the priority of
+   *     that page.
+   */
+  getPageDistanceMapHelper_(priority, map, pageId) {
+    if (map[pageId] !== undefined && map[pageId] <= priority) {
+      return map;
+    }
+
+    map[pageId] = priority;
+    const page = this.getPageById_(pageId);
+    page.getAdjacentPageIds().forEach(adjacentPageId => {
+      if (map[adjacentPageId] !== undefined
+          && map[adjacentPageId] <= priority) {
+        return;
+      }
+
+      map = this.getPageDistanceMapHelper_(priority + 1, map, adjacentPageId);
+    });
+
+    return map;
+  }
+
+
+  /** @private */
+  preloadPagesByDistance_() {
+    const pagesByDistance = this.getPagesByDistance_();
+
+    this.mutateElement(() => {
+      pagesByDistance.forEach((pageIds, distance) => {
+        pageIds.forEach(pageId => {
+          const page = this.getPageById_(pageId);
+          setStyles(page.element, {
+            transform: `translateY(${100 * distance}%)`,
+          });
+        });
+      });
+    });
+  }
+
+
+  /** @private */
+  buildBookend_() {
+    if (this.bookend_.isBuilt()) {
+      return Promise.resolve();
+    }
+
+    this.element.appendChild(this.bookend_.build());
+
+    this.setAsOwner(this.bookend_.getRoot());
+
+    return this.loadBookendConfig_().then(bookendConfig => {
+      if (bookendConfig !== null) {
+        this.bookend_.setConfig(dev().assert(bookendConfig));
+      }
+      this.scheduleResume(this.bookend_.getRoot());
+    });
+  }
+
+
+  /**
+   * @return {!Promise<?./bookend.BookendConfig>}
+   * @private
+   */
+  loadBookendConfigImpl_() {
+    // two-tiered implementation for backwards-compatibility with
+    // related-articles attribute
+    return this.loadBookendConfigInternal_().then(bookendConfig =>
+        bookendConfig || this.loadRelatedArticlesAsBookendConfig_());
+  }
+
+
+  /**
+   * @return {!Promise<?./bookend.BookendConfig>}
+   */
+  loadBookendConfigInternal_() {
+    return this.loadJsonFromAttribute_(BOOKEND_CONFIG_ATTRIBUTE_NAME)
+        .then(response => response && {
+          shareProviders: response['share-providers'],
+          relatedArticles: response['related-articles'] ?
+              buildFromJson(response['related-articles']) : [],
```

Comment:
Minor: Can we change `buildFromJson` to just return `[]` on undefined input?

Reply:
Filed #11549